### DISCUSSION TITLE
Challenge: Getting this IPC window switcher written in Python to work

### DISCUSSION THREAD

**Discussion by tobeypeters** [i3/i3#6128] [Q&A] [answered]
So, I wrote a window switcher, to cycle through all the windows on every workspace.  If I run it from the terminal manually, it works.  But if I use a keybinding or simply execute it from the i3 config, it runs but the you can switch apps.

cycle.py: https://github.com/tobeypeters/dotfiles/blob/master/.config/polybar/scripts/cycle.py
cycle.py helper: https://github.com/tobeypeters/dotfiles/blob/master/.config/polybar/scripts/my_helper.py
i3 config: https://github.com/tobeypeters/dotfiles/blob/master/.config/i3/config#L240

**Comment by tobeypeters** [comment]
I updated the code & it mostly works. But, it doesn't work with i3-msg:
`process_exec(f'i3-msg -q [id={windows[focus_idx]}] focus &') # focus it using i3-msg`

I had to switch it to:
`process_exec(f'xdotool windowactivate {windows[focus_idx]} &') # focus it using i3-msg`

So, any help would be appreicated.
I NEED it to use i3-msg.  xdotool isn't normally installed on peoples systems.  So, I don't want it to be a requirement to have installed.

**Reply by orestisfl** [reply, marked as answer]
Make sure to quote i3-msg's arguments. When working with scripts it's usually better to use exec() with program arguments directly instead of through a new shell. In i3's case it would be even better to use the IPC directly instead of i3-msg.

**Reply by tobeypeters** [reply]
I got it working, with IPC directly.

**Comment by orestisfl** [comment]
What did you try exactly?
